lib/Sema/TypeCheckExpr.cpp

```diff
@@ -655,116 +656,78 @@ static Type lookupDefaultLiteralType(TypeChecker &TC, DeclContext *dc,
-    type = &StringLiteralType;
-    name = "StringLiteralType";
-  }
-  // ExpressibleByIntegerLiteral -> IntegerLiteralType
-  else if (protocol == getProtocol(
-                         SourceLoc(),
-                         KnownProtocolKind::ExpressibleByIntegerLiteral)) {
-    type = &IntLiteralType;
-    name = "IntegerLiteralType";
-  }
-  // ExpressibleByFloatLiteral -> FloatLiteralType
-  else if (protocol == getProtocol(SourceLoc(),
-                                   KnownProtocolKind::ExpressibleByFloatLiteral)){
-    type = &FloatLiteralType;
-    name = "FloatLiteralType";
-  }
-  // ExpressibleByBooleanLiteral -> BoolLiteralType
-  else if (protocol == getProtocol(
-                         SourceLoc(),
-                         KnownProtocolKind::ExpressibleByBooleanLiteral)){
-    type = &BooleanLiteralType;
-    name = "BooleanLiteralType";
-  }
-  // ExpressibleByArrayLiteral -> Array
-  else if (protocol == getProtocol(SourceLoc(),
-                                   KnownProtocolKind::ExpressibleByArrayLiteral)){
-    type = &ArrayLiteralType;
-    name = "Array";
-    performLocalLookup = false;
-  }
-  // ExpressibleByDictionaryLiteral -> Dictionary
-  else if (protocol == getProtocol(
-                         SourceLoc(),
-                         KnownProtocolKind::ExpressibleByDictionaryLiteral)) {
-    type = &DictionaryLiteralType;
-    name = "Dictionary";
-    performLocalLookup = false;
-  }
-  // _ExpressibleByColorLiteral -> _ColorLiteralType
-  else if (protocol == getProtocol(
-                         SourceLoc(),
-                         KnownProtocolKind::ExpressibleByColorLiteral)) {
-    type = &ColorLiteralType;
-    name = "_ColorLiteralType";
-  }
-  // _ExpressibleByImageLiteral -> _ImageLiteralType
-  else if (protocol == getProtocol(
-                         SourceLoc(),
-                         KnownProtocolKind::ExpressibleByImageLiteral)) {
-    type = &ImageLiteralType;
-    name = "_ImageLiteralType";
-  }
-  // _ExpressibleByFileReferenceLiteral -> _FileReferenceLiteralType
-  else if (protocol == getProtocol(
-                         SourceLoc(),
-                         KnownProtocolKind::ExpressibleByFileReferenceLiteral)) {
-    type = &FileReferenceLiteralType;
-    name = "_FileReferenceLiteralType";
-  }
+  auto typeNameAndPerformLocalLookup =
+      getDefaultTypeNameAndPerformLocalLookup(protocol);
+  if (!typeNameAndPerformLocalLookup.first)
+    return nullptr;
+  Type type = evaluateOrDefault(
+      Context.evaluator,
+      DefaultTypeRequest{typeNameAndPerformLocalLookup.first,
+                         typeNameAndPerformLocalLookup.second, dc},
+      nullptr);
+  return type;
+}
 
-  if (!type)
+static std::pair<const char *, bool>
+getDefaultTypeNameAndPerformLocalLookup(const ProtocolDecl *protocol) {
+  TypeChecker &tc = TypeChecker::createForContext(protocol->getASTContext());
+
+#define TYPE_NAME_CASE(KIND, NAME, PERFORM_LOCAL_LOOKUP)                       \
```

Comment:
Very nice refactoring into a table, thank you!

Reply:
My pleasure! Thought about using a data structure instead of macros, but this way maintains the performance characteristics of what was there. Also the `std::array` constructor wanted a length, and I was loath to redundantly specify one. Although, if we move to an enum instead of `const char*`, hmmm....